### ISSUE TITLE
Impossible to use in ASM.JS

### ISSUE THREAD

**Issue by Subsentient** [rustwasm/wasm-bindgen#2483]
wasm-bindgen, as written, makes it impossible to convert to ASM.JS, as it leaves unresolved symbols, and the stub file generated assumes WebAssembly.

**Comment by CrimsonCodes0** [comment]
Well, the tool is dedicated toward Wasm, as it's even in the name.

Nothing here could be polyfilled in Asm.js anyway, all asm.js function imports must be for side-effects only, that is, they must have the signature `() -> ()`, which wasm-bindgen does not limit itself to.

If I may ask, out of personal interest, what did you intend to use asm.js for? What situation did you find yourself in where wasm is unsupported, yet asm.js was?

**Comment by Subsentient** [comment]
> If I may ask, out of personal interest, what did you intend to use asm.js for? What situation did you find yourself in where wasm is unsupported, yet asm.js is?

Sure. This was going to be for a small personal project of mine.

(Warzone 2100)[https://github.com/Warzone2100/warzone2100] has recently replaced its deprecated Qt4-based JavaScript engine with QuickJS in 4.x, which supports asm.js. Personally, I'm not a fan of JavaScript as a language, and so I was hoping to use Rust to write a new AI for the game. Obviously an embedded interpreted-only JavaScript engine like this does not support WebAssembly, so you see my use-case.

**Comment by CrimsonCodes0** [comment]
Sorry, I should've been clearer, by "asm.js support," I was referring to asm.js support, as seperate from "normal" JavaScript support, e.g. an AoT compiler dedicated to asm.js code.

As having used QuickJS before, it doesn't support either.

But, I guess that is one of the key parts of asm.js: compatibility with older engines.

**Comment by Subsentient** [comment]
> Sorry, I should've been clearer, by "asm.js support," I was referring to asm.js support, as seperate from "normal" JavaScript support, e.g. an AoT compiler dedicated to asm.js code.
> 
> As having used QuickJS before, it doesn't support either.
> 
> But, I guess that is one of the key parts of asm.js: compatibility with older engines.

Ahh, no I should have been more clear. Yeah QuickJS has no dedicated fast paths for asm.js, but it does execute asm.js correctly, which was something the old engine was incapable of doing.

**Comment by RReverser** [comment]
Do you really need **asm.**js for your use-case though? Since there's no dedicated fast path, you can just as well use `wasm2js` from the Binaryen toolchain to transpile the resulting Wasm: https://github.com/WebAssembly/binaryen

**Comment by CrimsonCodes0** [comment]
Had you checked out https://rustwasm.github.io/docs/wasm-bindgen/examples/wasm2js.html?

Are you suggesting that the steps there had failed?

**Comment by skelewir** [comment]
Yes, the steps from https://rustwasm.github.io/docs/wasm-bindgen/examples/wasm2js.html do fail :(

**Comment by Liamolucko** [comment]
> Yes, the steps from https://rustwasm.github.io/docs/wasm-bindgen/examples/wasm2js.html do fail :(

Fail how? Is there an error message?
